### ISSUE TITLE
Feature Request: Review platform

### ISSUE THREAD

**Issue by Nosferican** [JuliaComputing/JuliaHub#2]
It would be great for allowing users (e.g., registered users) to rate packages (e.g., 👍/ 👎, I use this, etc.).

**Comment by DilumAluthge** [comment]
In particular, it would be nice to have a way of doing peer code review. For example, I can do code review on a package and post my review.

It would also be good to divide the review into "security-related" and "non-security-related". The "non-security-related" part would be posted publicly. The "security-related" part would only be shared with the owners of the package.

For example, if I suggested updating a docstring, or refactoring a method, that would be "non-security-related", and that feedback could be shared publicly.

In contrast, if I identified a security vulnerability in a package, it would be irresponsible to share that publicly before the owners of the package had a chance to fix it. So that would not be shared publicly.

**Comment by ViralBShah** [comment]
Isn't a lot of this replicating what Github already provides?

**Comment by DilumAluthge** [comment]
> Isn't a lot of this replicating what Github already provides?

GitHub provides a way to do pull request reviews, but I'm not aware of a method for doing reviews on an entire repository.

I also don't think you can make any part of a pull request review private.

It's possible, however, that this functionality doesn't necessarily need to live in JuliaHub.

**Comment by Nosferican** [comment]
Every GitHub repository has the Security tab which allows for a policy and secure reporting of vulnerabilities. There are also a bunch of integrations with https://securitylab.github.com/ and security alarms.

What it doesn't provide much is for user reviews à la https://www.crantastic.org/packages/ggplot or or such. Right now GitHub only supports the staring the repository. Reviews could indicate when to use one package vs the other, provide benchmarks, etc.

**Comment by DilumAluthge** [comment]
I like the idea of user reviews, i.e. people write a paragraph or two about the package.

I don't like the idea of numerical ratings like on crantastic, where people rate a package on a scale of 0 to 5 or 1 to 5 or something like that. I don't think that contributes anything.

**Comment by KristofferC** [comment]
Adding comments comes with a whole lot of baggage. You need moderation, prevent people brigading, deal with spam. Just looking at the crantastic link, all reviews I see are spam?

<img width="406" alt="image" src="https://user-images.githubusercontent.com/1282691/81740699-a9eed780-949d-11ea-9a21-30fc778a67e0.png">

Personally, I don't think this is functionality that is a good fit for JuliaHub

**Comment by Nosferican** [comment]
Maybe the GitHub stars, reverse dependencies, and a 5 stars rate for documentations might be enough. I would imagine that posting a review would require being registered but maybe is not as worthwhile as a more basic low cost system.

**Comment by c42f** [comment]
Regarding unsolicited whole-package reviews, I do think they're useful, but I've posted these as github issues before: largely they end up being very high-level; so less about the detail of the code and more architectural. This seemed to work fine.

For package authors who are explicitly soliciting feedback, they can
* Get that on discourse, if they want informal review
* Ask collaborators to post it in a github issue
* Publish to a peer-reviewed software journal like JOSS (and many others https://www.software.ac.uk/which-journals-should-i-publish-my-software)

All these are good existing options and I think they cover a fairly wide variety of use cases without JuliaHub getting into the messy business of comment moderation. I imagine we could provide links to articles and reviews hosted elsewhere though.

**Comment by DilumAluthge** [comment]
I think the best way forward for whole-package code review will be implementing support for `crev` (https://github.com/crev-dev/crev) in Julia/Pkg/General.

See for example the `crev` integration with Rust/Cargo: https://github.com/crev-dev/cargo-crev

**Comment by Nosferican** [comment]
@non-Jedi, could you give a roadmap of what would it entail to get `crev` support for Julia?

**Comment by non-Jedi** [comment]
Sure. From discussions with @dpc (main Crev author I'm tagging in case he has anything interesting to add; feel free to unsubscribe) on the [crev matrix channel](https://matrix.to/#/!uBhYhtcoNlyEbzfYAW:matrix.org?via=matrix.org&via=thebeckmeyers.xyz&via=pixie.town), I think the best way forward is to re-implement the crev format and logic in Julia. The code for [the Rust implementation](https://github.com/crev-dev/cargo-crev) (the only currently usable implementation) is pretty tied to Rust/cargo and would require quite a bit of rework to allow it to be used as a shared library or a command-line program Julia could shell out to. That said, if someone who wants to see this feature is fluent in Rust, doing that work may be faster than reimplementing in Julia.

My best 5-minute effort at a roadmap (with the few things I've completed at https://github.com/non-Jedi/Crev.jl) checked off would look something like:

- [X] Parse and verify signatures of Crev proofs.
- [ ] Implement package digest hashing or get [the reference implementation](https://github.com/crev-dev/rblake2sum) building with BinaryBuilder.
- [ ] Generate and sign a proof of a directory.
- [ ] Hook into `Pkg` for doing proofs of Julia packages and returning which package a proof belongs to.
- [ ] Parse and verify "trust" proofs.
- [ ] Generate trust proofs.
- [ ] Implement transitive trust (web-of-trust).
- [ ] Build frontend as extension of `Pkg` REPL mode or as a separate REPL mode.
- [ ] Drive ecosystem adoption (cargo-crev is not super widely used within the Rust ecosystem).

I'm sure I'm missing some steps that need implemented, but I think that gives a rough idea of the complexity here.

**Comment by Nosferican** [comment]
That seems like a good case for testing out the Pkg extensions model.

**Comment by dpc** [comment]
@non-Jedi Agreed. The author of [pip/Python frontend](https://github.com/crev-dev/pip-crev) is going with the "use Rust crev-lib as a library" route - there are nice crates that make it easy to create Rust-Python bindings though. But at least the effort to keep the library in the right shape for binding authors is already there.

I tried to make things straightforward, so that it's not difficult to implement everything from scratch. Crypto is just ed25519, proofs are yaml + a bit of manual envelope parsing, WoT algorithm is provided as a reference, and implementations might chose their own variants. The big advantage of this route is that it's easier for Julia community to work and improve Julia codebase.

One way or another it shouldn't be too hard. I'm happy to help answer any questions as they arise.

**Comment by c42f** [comment]
It seems there may be some overlap here with the new "curation and trust" issue Stefan just opened at https://github.com/JuliaLang/Pkg.jl/issues/1856

**Comment by tkf** [comment]
I tried to figure out how crev identifies a release.  Looking at [the example crev package review](https://github.com/dpc/crev-proofs/blob/2d250e26bed95927a76551c7969cd108ebb1946c/FYlr8YoYGVvDwHQxqEIs89reKKDy-oWisoO0qXXEfHE/reviews/2018-12-packages-Ua7DxQ.proof.crev#L84), is this the information used for identifying a release of a package?

```yaml
package:
  source: "https://crates.io"
  name: either
  version: 1.5.0
  digest: uBbgCVotv_8z4SEOjremFmvMG4JPhUROC19OLjPPLNE
```

I guess `source:` can be the URL of registry?  It's easy to fill `name:` and `version:` (and I suppose digest is https://github.com/crev-dev/recursive-digest).  But this is not enough for identifying a Julia package.  Is it possible to have `uuid:` or something in `package:`?  Otherwise, maybe we can use something like `https://pkg.julialang.org/package/$uuid` as `source:`?

**Comment by DilumAluthge** [comment]
I think two things have become clear:
1. JuliaHub is not the right place for this.
2. Adding Julia support for `crev` and integrating `crev` into Pkg is probably going to be the best path forward

Perhaps we can close this issue?

For integrating `crev` into Pkg, I've opened this meta-issue: https://github.com/JuliaLang/Pkg.jl/issues/1858

**Comment by Nosferican** [comment]
Registered packages are uniquely identified by the UUID. Versions have a semver tag / name and are identified by the Git Tree hash.

**Comment by dpc** [comment]
@tkf `source` is meant to identify the registry & language. Basically - a namespace in which package names are supposed to be unique. The details are not very well defined yet because it's hard to make any definitive decisions without more examples and uses-cases. In Rust, which is my primary target, `source` is used to support alternative registries (as opposed to default crates.io).

About uuids: `source + name` is supposed to uniquely identify the package in question. One could either make `name: packagename#uuid` or something which I guess is readable enough, or crev / Julia's implementation of it, could just add an additional field `uuid` or something like that, and we can make `source + name + uuid` to be unique, where `uuid` is optional.

Yaml was specifically chosen with an idea that it's easy to extend, and implementations should ignore fields they don't understand. And the everything is still fresh and easy to change.

**Comment by tkf** [comment]
@dpc Thanks for the answers!

> In Rust, which is my primary target, `source` is used to support alternative registries (as opposed to default crates.io).

Julia has a similar concept.  So, I guess `source: "https://github.com/JuliaRegistries/General"` would be a natural choice.

> About uuids: `source + name` is supposed to uniquely identify the package in question. One could either make `name: packagename#uuid` or something which I guess is readable enough, or crev / Julia's implementation of it, could just add an additional field `uuid` or something like that, and we can make `source + name + uuid` to be unique, where `uuid` is optional.

Unfortunately, `source + name` may not uniquely identify the package in Julia.  So including the UUID in `name` sounds like a good option.  We use `packagename=uuid` in the package manager UI so I guess it'd be a good candidate.  Adding `uuid` might be cleaner but let's see how other people think about this.

So, concretely, I guess the format would be something like

```yaml
package:
  source: "https://github.com/JuliaRegistries/General"
  name: Example=7876af07-990d-54b4-ab0e-23690620f79a
  version: 0.5.4
```

or

```yaml
package:
  source: "https://github.com/JuliaRegistries/General"
  name: Example
  uuid: 7876af07-990d-54b4-ab0e-23690620f79a
  version: 0.5.4
```
